data/blog/2025-02-06-hacker-news-x-agent.mdx

```diff
@@ -0,0 +1,367 @@
+```
+
+### Scraping the Story
+
+Given the URL of the story by Hacker News, we will scrape the content of the story using `cheerio`.
+
+1. Fetch the HTML content of the story using `fetch`.
+2. Load the HTML content into `cheerio`, a library for parsing HTML.
+3. Remove unwanted elements from the content and extract the main content.
+4. Clean the content by removing extra spaces and newlines.
+
+```typescript
+import * as cheerio from "cheerio";
+import { SELECTORS_TO_REMOVE } from "@/app/constants";
+
+if (!url) {
+  return {
+    title,
+    url,
+    content: "",
+  };
+}
+
+const html = await fetch(url).then((res) => res.text());
+
+const $ = cheerio.load(html);
+
+SELECTORS_TO_REMOVE.forEach((selector) => {
+  $(selector).remove();
+});
+
+let $content = $('main, article, [role="main"]');
+
+if (!$content.length) {
+  $content = $("body");
+}
+
+const content = $content
+  .text()
+  .replace(/\s+/g, " ")
+  .replace(/\n\s*/g, "\n")
+  .trim();
+
+return {
+  title,
+  url,
+  content,
+};
+```
+
+### Creating the `hackerNewsTool`
+
+To provide the functionality we developed to our agent, we will create an AI SDK compatible tool, specifying the description, parameters, and execution function.
+
+```typescript
+import { z } from "zod";
+import { tool } from "ai";
+
+tool({
+  description:
+    "A tool for fetching the top 1 unvisited Hacker News article. It returns an " +
+    "object with the title, url, and content of the article. It does not take any " +
+    "parameters, so give an empty object as a parameter. You absolutely should not " +
+    "give an empty string directly as a parameter.",
+  parameters: z.object({}),
+  execute: async ({}) => {
+    // Fetching the Top Stories
+    // Picking an Unseen Story
+    // Scraping the Story
+  },
+});
+```
+
+## Tweeting the Story
+
+### Generating an image
+
+We will use `context.call` to call the Ideogram API. `context.call` enables us to perform HTTP calls as a workflow step. We won't need to actively wait for the call to complete, preventing function timeouts and reducing costs. A cool trick we are doing here is passing `DESIGN` as the style type and using a color palette to get a more consistent style across the images.
+
+```typescript
+const { body: ideogramResult } = (await context.call(
+  "call image generation API",
+  {
+    url: "https://api.ideogram.ai/generate",
+    method: "POST",
+    body: {
+      image_request: {
+        model: "V_2",
+        prompt: imagePrompt,
+        aspect_ratio: "ASPECT_16_9",
+        magic_prompt_option: "AUTO",
+        style_type: "DESIGN",
+        color_palette: {
+          members: [
+            { color_hex: "#FF6D00" },
+            { color_hex: "#FFCA12" },
+            { color_hex: "#58BAE7" },
+            { color_hex: "#DDDDDD" },
+          ],
+        },
+      },
+    },
+    headers: {
+      "Content-Type": "application/json",
+      "Api-Key": process.env.IDEOGRAM_API_KEY!,
+    },
+  }
+)) as { body: IdeogramResponse };
```

Comment:
we don't need to do `as { body: IdeogramResponse }`. You should be able to pass `context.call<IdeogramResponse>`. We can fix this in the code too.